src/test/java/org/xbill/DNS/DNSSessionTest.java

```diff
@@ -0,0 +1,225 @@
+    verifyNoMoreInteractions(mockResolver);
+  }
+
+  @Test
+  public void lookupAsync_absoluteQuery() throws InterruptedException, ExecutionException {
+    wireUpMockResolver(mockResolver, query -> answer(query, name -> LOOPBACK_A));
+
+    DNSSession dnsSession = new DNSSession(mockResolver);
+    CompletionStage<LookupResult> resultFuture =
+        dnsSession.lookupAsync(Name.fromConstantString("a.b."), A, IN);
+
+    LookupResult result = resultFuture.toCompletableFuture().get();
+    assertEquals(singletonList(LOOPBACK_A.withName(name("a.b."))), result.get());
+
+    verify(mockResolver).sendAsync(any());
+  }
+
+  @Test
+  public void lookupAsync_simpleCnameRedirect() throws Exception {
+    Function<Name, Record> nameToRecord =
+        name -> name("cname.r.").equals(name) ? cname("cname.r.", "a.b.") : LOOPBACK_A;
+    wireUpMockResolver(mockResolver, q -> answer(q, nameToRecord));
+
+    DNSSession dnsSession = new DNSSession(mockResolver);
+
+    CompletionStage<LookupResult> resultFuture = dnsSession.lookupAsync(name("cname.r."), A, IN);
+
+    LookupResult result = resultFuture.toCompletableFuture().get();
+    assertEquals(singletonList(LOOPBACK_A.withName(name("a.b."))), result.get());
+    verify(mockResolver, times(2)).sendAsync(any());
+  }
+
+  @Test
+  public void lookupAsync_simpleDnameRedirect() throws Exception {
+    Function<Name, Record> nameToRecord =
+        n -> name("x.y.to.dname.").equals(n) ? dname("to.dname.", "to.a.") : LOOPBACK_A;
+    wireUpMockResolver(mockResolver, q -> answer(q, nameToRecord));
+
+    DNSSession dnsSession = new DNSSession(mockResolver);
+
+    CompletionStage<LookupResult> resultFuture =
+        dnsSession.lookupAsync(name("x.y.to.dname."), A, IN);
+
+    LookupResult result = resultFuture.toCompletableFuture().get();
+    assertEquals(singletonList(LOOPBACK_A.withName(name("x.y.to.a."))), result.get());
+    verify(mockResolver, times(2)).sendAsync(any());
+  }
+
+  @Test
+  public void lookupAsync_redirectLoop() {
+    Function<Name, Record> nameToRecord =
+        name -> name("a.b.").equals(name) ? cname("a.", "b.") : cname("b.", "a.");
+    wireUpMockResolver(mockResolver, q -> answer(q, nameToRecord));
+
+    DNSSession dnsSession = new DNSSession(mockResolver);
+    dnsSession.setMaxRedirects(2);
+
+    CompletionStage<LookupResult> resultFuture =
+        dnsSession.lookupAsync(name("first.example.com."), A, IN);
+
+    assertThrowsCause(
+        RedirectOverflowException.class, () -> resultFuture.toCompletableFuture().get());
+    verify(mockResolver, times(3)).sendAsync(any());
+  }
+
+  @Test
+  public void lookupAsync_NXDOMAIN() throws Exception {
+    wireUpMockResolver(mockResolver, q -> fail(q, Rcode.NXDOMAIN));
+
+    DNSSession dnsSession = new DNSSession(mockResolver);
+    CompletionStage<LookupResult> resultFuture = dnsSession.lookupAsync(name("a.b."), A, IN);
+
+    assertEquals(new LookupResult(emptyList(), NXDOMAIN), resultFuture.toCompletableFuture().get());
```

Comment:
I'm just not convinced that an empy list is really the best answer for NXDOMAIN, but neither are `Optional`s or exceptions.

Ideally the return type would be tuple of (result, records) so callers would be forced to not ignore it.

Reply:
Do you have an idea/examples of how non-unit test sync and async code should look like?

Reply:
Let me think some more about how calling code might look and get back to you on that.

Reply:
I'll write something as a new comment below, as this is a comment on an outdated piece of code.